core/src/main/scala/pureconfig/ConfigCursor.scala

```diff
@@ -66,7 +66,33 @@ sealed trait ConfigCursor {
    *         otherwise.
    */
   def asListCursor: Either[ConfigReaderFailures, ConfigListCursor] =
-    castOrFail(LIST, Set(LIST), _.asInstanceOf[ConfigList]).right.map(ConfigListCursor(_, pathElems))
+    if (isUndefined) {
+      failed(KeyNotFound.forKeys(path, Set()))
+    } else {
+      val ve = value.valueType() match {
+        case ConfigValueType.OBJECT =>
+          val ll =
+            value.asInstanceOf[ConfigObject].asScala.foldLeft[Either[List[String], List[(Int, ConfigValue)]]](Right(Nil)) {
+              case (acc, (str, v)) =>
+                (acc, Try(str.toInt)) match {
+                  case (Right(a), Success(b)) => Right(a :+ (b -> v))
+                  case (Left(keys), Failure(_)) => Left(keys :+ str)
+                  case (_, Failure(_)) => Left(List(str))
+                  case (Left(keys), _) => Left(keys)
+                }
+            }
+
+          ll.left.map(CannotConvertObjectToList)
```

Comment:
Should probably sort the keys alphabetically for nice/repeatable error message.

Reply:
Makes sense. I've sorted them alphabetically in ee993eb969ef01a53b1603ae96d8a5e07d80af7b.